### PR TITLE
#164198145 Add support for social authentication

### PR DESCRIPTION
**Pull request by devdbrandy**
#### What does this PR do?

This provides support for social authentication (OAuth) with facebook and google.

#### Description of Task to be completed

- [x] Provide social buttons
- [x] Add action to be dispatched upon request
- [x] Add function to handle request
- [x] Add function to handle request failures
- [x] Add isAuthenticated global state
- [x] Provide additional global states
- [x] Configure dotenv with webpack
- [x] Provide quick helper functions
- [x] Write test against implementation

#### How should this be manually tested

1. Clone repo by running the following command on your terminal:

```bash
git clone -b feature/164198145/social-authentication --single-branch \
https://github.com/andela/apollo-ah-frontend.git
```

2. Navigate into the repo directory: `cd apollo-ah-frontend`

2. Install dependency packages with: `npm install`

3. On the root directory, create a `.env` file from `.env.example` and provide required credentials

4. Run `npm test`

#### Any background context you want to provide

Enhancement has been provided for the Web API platform: [PR #60](https://github.com/andela/apollo-ah-backend/pull/60)

#### What are the relevant pivotal tracker stories

[#164198145](https://www.pivotaltracker.com/story/show/164198145)

#### Screenshots

social-auth-screencast-v3

### CODE REVIEW

File: src/components/HomePage.jsx [hunk 1]
```diff
@@ -9,8 +9,8 @@ class HomePage extends React.Component {
   render() {
     return (
       <div>
-        <p>Hello World</p>
-        <a Link="/login" href="/login">Login </a>
+        <p>Home Page</p>
+        <Link to="/login">Login</Link>
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply src/components/HomePage.jsx]: `Link` component generates href attribute (anchor tag), which is valid from the accessibility point of view. Implemented resolve: https://github.com/evcohen/eslint-plugin-jsx-a11y/issues/340#issuecomment-338424908

File: src/actions/socialAction.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          'Access-Control-Allow-Origin': '*',
+        },
+      body: JSON.stringify(payload)
+    };
+
+    return fetch(`${process.env.API_BASE_URL}/auth/social`, requestOptions)
+      .then(res => res.json())
```
Comment (Andhrah) [inline src/actions/socialAction.js]: Good job @devdbrandy . Kindly make use of async/await here.

**Review by Andhrah — COMMENTED** [review]
Nice job @devdbrandy, your code was readable at first glance. Kindly attend to the requested changes

File: src/actions/socialAction.js (hunk 2 above)
Comment (devdbrandy) [reply src/actions/socialAction.js]: Hi @Andraquin, thanks for the feedback. I have implemented the requested changes to maintain readability.

**Review by Andhrah — APPROVED** [review]
Good job @devdbrandy.

**Review by Andhrah — APPROVED** [review]
Great work @devdbrandy

File: src/actions/globalActions.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import {
```
Comment (aanchirinah) [inline src/actions/globalActions.js]: #### Nitpick
- Use selectors

File: src/components/SocialLogin.jsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    isAuthenticated: PropType.bool.isRequired,
+  }
+  
+  /**
+   * Creates an instance of SocialLogin.
+   * 
+   * @returns {void}
+   * @memberof SocialLogin
+   */
+  constructor() {
```
Comment (aanchirinah) [inline src/components/SocialLogin.jsx]: #### Nitpick
- Why is there a constructor if we are not using it?

File: src/index.jsx [hunk 5]
```diff
@@ -1,23 +1,29 @@
...
 
-
-
 function App() {
   return (
-    <div>
+    <Provider store={store}>
       <Routes />
-    </div>
+      <ReduxToastr
```
Comment (aanchirinah) [inline src/index.jsx]: #### Nitpick
- Why ?

File: src/reducers/globalReducer.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import initialStore from '../store/initialState';
+import {
+  TRIGGER_AUTHENTICATED,
+  TRIGGER_ERROR,
+} from '../actions/actionTypes';
+
+const { global } = initialStore;
```
Comment (aanchirinah) [inline src/reducers/globalReducer.js]: #### Nitpick 
- Why the choice of name `globalReducer` for this file?

File: src/reducers/globalReducer.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Glabal related state handler
+ *
+ * @export
+ * @param {object} [state=global] - The global state object
+ * @param {object} action - The global action object
+ * @returns {object} The global state object
+ */
+export default function globalReducer(state = global, action) {
+  if (action.type === TRIGGER_AUTHENTICATED) {
```
Comment (aanchirinah) [inline src/reducers/globalReducer.js]: #### Nitpick
- `TRIGGER_AUTHENTICATED` why not `AUTHENTICATION_SUCCESS`/`AUTHENTICATION_FAILURE`

File: src/views/common/GoogleButton.jsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import GoogleLogin from 'react-google-login';
```
Comment (aanchirinah) [inline src/views/common/GoogleButton.jsx]: #### Nitpick
- Is this really necessary?

**Review by aanchirinah — CHANGES_REQUESTED** [review]
I still have lots of questions but good job, kindly attend to the comments and requested changes.

File: src/components/SocialLogin.jsx (hunk 4 above)
Comment (devdbrandy) [reply src/components/SocialLogin.jsx]: Thanks for pointing that. Initially the constructor was used to bind class methods. However, I realise it's redundant.

File: src/index.jsx (hunk 5 above)
Comment (devdbrandy) [reply src/index.jsx]: The plugin was initially used for demonstration purpose. However, I will make an update of a recommended toaster with support for easy customising: [react-toastify](https://github.com/fkhadra/react-toastify)

File: src/reducers/globalReducer.js (hunk 6 above)
Comment (devdbrandy) [reply src/reducers/globalReducer.js]: It is generic to the `global` state declared in the app store.

*initialStore.js*
Screen Shot 2019-04-10 at 9 28 12 AM

File: src/reducers/globalReducer.js (hunk 7 above)
Comment (devdbrandy) [reply src/reducers/globalReducer.js]: Thanks for pointing that, I think `AUTHENTICATION_SUCCESS` sounds good. I will consider refactoring.

File: src/views/common/GoogleButton.jsx (hunk 8 above)
Comment (devdbrandy) [reply src/views/common/GoogleButton.jsx]: Doing more research on the implementation, I've come up with a solution that won't require us using additional plugin. I will also make an update regarding that.

File: src/reducers/globalReducer.js (hunk 6 above)
Comment (aanchirinah) [reply src/reducers/globalReducer.js]: @devdbrandy , What is this `global` state? You know `global` is a reserved word in lots of coding languages right?

File: src/actions/globalActions.js (hunk 3 above)
Comment (devdbrandy) [reply src/actions/globalActions.js]: Resolved

File: src/reducers/globalReducer.js (hunk 6 above)
Comment (devdbrandy) [reply src/reducers/globalReducer.js]: Yes, I'm aware. However, I have made changes to the initial state.

File: src/utils/helpers.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import jwt from 'jsonwebtoken';
+
+/**
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply src/utils/helpers.js:3]: Resolved

File: src/components/SocialLogin.jsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * Called immediately before mounting occurs
+   *
+   * @returns {void}
+   * @memberof SocialLogin
+   */
+  componentWillMount() {
+    const { socialLogin } = this.props;
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply src/components/SocialLogin.jsx]: Resolved

File: src/actions/socialAction.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    return dispatch(authenticationSuccess(payload));
+  } catch (error) {
+    return dispatch(authenticationFailure(error.message));
+  }
+};
+
+/**
+ * Triggers authentication success
+ *
+ * @param {boolean} isAuthenticated - The user authenticated state
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply src/actions/socialAction.js]: Resolved

**Review by valenstical — CHANGES_REQUESTED** [review]
@devdbrandy, you may need to resolve merge conflicts. Great work!

**Comment by devdbrandy** [comment]
Hi @valenstical, thanks for pointing that. I have resolved the merge conflicts.

**Review by valenstical — APPROVED** [review]
LGTM

**Comment by devdbrandy** [comment]
Hi @aanchirinah, I have commented out the twitter button component (commit 2661afd) because it requires more stress testing.

**Review by aanchirinah — APPROVED** [review]
LGTM

**Pull request merged**